graph-builder/src/registry.rs

```diff
@@ -114,21 +116,67 @@ pub fn fetch_releases(
         ))
         .map_err(|e| format_err!("{}", e))?;
 
-    let releases = get_tags(repo.to_owned(), authenticated_client)
+    let tags = get_tags(repo.to_owned(), authenticated_client)
         .and_then(|tag| get_manifest_and_layers(tag, repo.to_owned(), authenticated_client))
-        .and_then(|(tag, layer_digests)| {
-            find_first_release(
-                layer_digests,
-                authenticated_client.to_owned(),
-                registry_host.to_owned(),
-                repo.to_owned(),
-                tag.to_owned(),
-            )
-        })
-        .filter_map(|release| release)
         .collect();
+    let tagged_layers = thread_runtime.block_on(tags)?;
+
+    let mut releases = Vec::with_capacity(tagged_layers.len());
+    for (tag, layer_digests) in tagged_layers {
+        let release = cache_release(
+            layer_digests,
+            authenticated_client.to_owned(),
+            registry_host.to_owned(),
+            repo.to_owned(),
+            tag.to_owned(),
+            cache,
+        )?;
+        if let Some(metadata) = release {
+            releases.push(metadata);
+        };
+    }
+    releases.shrink_to_fit();
 
-    thread_runtime.block_on(releases)
+    Ok(releases)
+}
+
+/// Look up release metadata for a specific tag, and cache it.
+///
+/// Tags and release metadata are immutable, thus each
+/// tagged release is looked up at most once and both
+/// positive and negative results cached indefinitely.
+fn cache_release(
+    layer_digests: Vec<String>,
+    authenticated_client: dkregistry::v2::Client,
+    registry_host: String,
+    repo: String,
+    tag: String,
+    cache: &mut HashMap<String, Option<Release>>,
+) -> Fallible<Option<Release>> {
+    // TODO(lucab): get rid of this synchronous lookup, by
+    // introducing a dedicated actor which owns the cache
+    // and handles queries and insertions.
+    let mut thread_runtime = tokio::runtime::current_thread::Runtime::new()?;
+
+    if let Some(release) = cache.get(&tag) {
+        trace!("Using cached release metadata for tag {}", &tag);
+        return Ok(release.clone());
+    }
+
+    let tagged_release = find_first_release(
+        layer_digests,
+        authenticated_client,
+        registry_host,
+        repo,
+        tag,
+    );
+    let (tag, release) = thread_runtime
+        .block_on(tagged_release)
+        .context("failed to find first release")?;
+
+    trace!("Caching release metadata for new tag {}", &tag);
+    cache.insert(tag, release.clone());
```

Comment:
If I didn't misread the `find_first_release` logic, a negative result gets here as a `None` release which is then cached for that tag.

Reply:
I will add a docstring on the data-structure to make this more explicit.

Reply:
I see, that makes sense. I missed that `cache.get(..)` wraps it so we're getting `Option<Option<Release>>`.